lara-typescript/src/interactive-api-client/hooks.ts

```diff
@@ -0,0 +1,86 @@
+import { useEffect, useState } from "react";
+import { IInitInteractive } from "./types";
+import * as client from "./api";
+
+export const useInteractiveState = <InteractiveState>() => {
+  const [ interactiveState, setInteractiveState ] = useState<InteractiveState | null>(client.getInteractiveState<InteractiveState>());
+
+  useEffect(() => {
+    // Setup client event listeners. They will ensure that another instance of this hook (or anything else
+    // using client directly) makes changes to interactive state, this hook will receive these changes.
+    const handleStateUpdate = (newState: InteractiveState) => {
+      setInteractiveState(newState);
```

Comment:
If I understand this right, the newState passed by the the client could be the same object just with different properties.  So then React will use `Object.is` and not think the object has changed. So then it won't trigger a re-render.

If the updateState approach was used above then this would be turned into:
`updateState(prev => prev + 1)`

Reply:
> So then React will use Object.is and not think the object has changed. So then it won't trigger a re-render.

Right, although I think this is developer's fault and personally I wouldn't try to fix it. The same way React doesn't try to fix it in `useState`. That's why I was using `Object.assign` in the interactives I was writing (even before this hook was added). But if others think that super-bulletproof approach is better, let me know.

Reply:
I think it is fine to keep it as is, but perhaps a comment in the client's 'setInteractiveState' definition about it would be useful. Since the handleStateUpdate call might be triggered by someone using the client directly to send a new state to LARA. If they aren't in the React part of the code, they might not realize that they need to copy the state in order to cause any React listeners to update.

Reply:
Good point, thanks. I've added comments. I've also added "freezing" of this state in the last comment + Readonly TypeScript modifier. So, it should be impossible to change this state.